test/e2e-test/appconfig_dependency_test.go

```diff
@@ -0,0 +1,475 @@
+
+	"github.com/crossplane/oam-kubernetes-runtime/apis/core/v1alpha2"
+	"github.com/crossplane/oam-kubernetes-runtime/pkg/oam/util"
+)
+
+var _ = Describe("Resource Dependency in an ApplicationConfiguration", func() {
+	ctx := context.Background()
+	namespace := "appconfig-dependency-test"
+	var ns corev1.Namespace
+	var wd v1alpha2.WorkloadDefinition
+	var td v1alpha2.TraitDefinition
+	tempFoo := &unstructured.Unstructured{}
+	tempFoo.SetAPIVersion("example.com/v1")
+	tempFoo.SetKind("Foo")
+	tempFoo.SetNamespace(namespace)
+	BeforeEach(func() {
+		ns = corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
+		logf.Log.Info("Start to run a test, clean up previous resources")
+		// delete the namespace with all its resources
+		Expect(k8sClient.Delete(ctx, &ns, client.PropagationPolicy(metav1.DeletePropagationForeground))).
+			Should(SatisfyAny(BeNil(), &util.NotFoundMatcher{}))
+		logf.Log.Info("make sure all the resources are removed")
+		objectKey := client.ObjectKey{
+			Name: namespace,
+		}
+		res := &corev1.Namespace{}
+		Eventually(
+			// gomega has a bug that can't take nil as the actual input, so has to make it a func
+			func() error {
+				return k8sClient.Get(ctx, objectKey, res)
+			},
+			time.Second*30, time.Millisecond*500).Should(&util.NotFoundMatcher{})
+		// recreate the namespace for testing
+		Eventually(
+			func() error {
+				return k8sClient.Create(ctx, &ns)
+			},
+			time.Second*3, time.Millisecond*300).Should(SatisfyAny(BeNil(), &util.AlreadyExistMatcher{}))
+		wd = v1alpha2.WorkloadDefinition{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "foo.example.com",
+			},
+			Spec: v1alpha2.WorkloadDefinitionSpec{
+				Reference: v1alpha2.DefinitionReference{
+					Name: "foo.example.com",
+				},
+			},
+		}
+		td = v1alpha2.TraitDefinition{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "foos.example.com",
+			},
+			Spec: v1alpha2.TraitDefinitionSpec{
+				Reference: v1alpha2.DefinitionReference{
+					Name: "foos.example.com",
+				},
+			},
+		}
+		logf.Log.Info("Creating workload definition")
+		// For some reason, WorkloadDefinition is created as a Cluster scope object
+		Expect(k8sClient.Create(ctx, &wd)).Should(SatisfyAny(BeNil(), &util.AlreadyExistMatcher{}))
+		logf.Log.Info("Creating trait definition")
+		// For some reason, TraitDefinition is created as a Cluster scope object
+		Expect(k8sClient.Create(ctx, &td)).Should(SatisfyAny(BeNil(), &util.AlreadyExistMatcher{}))
+	})
+	AfterEach(func() {
+		logf.Log.Info("Clean up resources")
+		// Delete the WorkloadDefinition and TraitDefinition
+		Expect(k8sClient.Delete(ctx, &wd)).Should(BeNil())
+		Expect(k8sClient.Delete(ctx, &td)).Should(BeNil())
+		// delete the namespace with all its resources
+		Expect(k8sClient.Delete(ctx, &ns, client.PropagationPolicy(metav1.DeletePropagationForeground))).Should(BeNil())
+	})
+
+	// common function for verification
+	verify := func(appConfigName, outName, inName string) {
```

Comment:
This is a good place to use table based testing style

Reply:
Ginkgo promotes BDD style and this is the correct way to do it.
AFAIK, TBT should not be used here otherwise the two styles would just conflict.

Reply:
I don't think Ginkgo conflicts with TBT but I am ok with either style, just not to copy-paste the same setup code again and again.

Reply:
>  just not to copy-paste the same setup code again and again.

Totally agree.